### PR TITLE
feat(auth): add options to resendSignUpCode

### PR DESCRIPTION
**Pull request by Jordan-Nelson**
*Issue #, if available:* n/a

*Description of changes:*
- Add ResendSignUpCodeOptions/AWSCognitoResendSignUpCodeOptions to enable passing metadata in resendSignUpCode()

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/options/AWSCognitoAuthResendSignUpCodeOptions.java [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+     * @return a builder object.
+     */
+    @NonNull
+    public static CognitoBuilder builder() {
+        return new CognitoBuilder();
+    }
+
+    @Override
+    public int hashCode() {
+        return ObjectsCompat.hash(
```
Comment (richardmcclellan) [inline aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/options/AWSCognitoAuthResendSignUpCodeOptions.java:64]: `hashCode`, `equals` should have calls to `super()`, right?

Comment (Jordan-Nelson) [reply aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/options/AWSCognitoAuthResendSignUpCodeOptions.java:64]: I don't think so. AuthResendSignUpCodeOptions is an abstract class and doesn't implement hashCode/equals. This is the pattern used for other similar options. Let me know if I am missing something.

File: rxbindings/src/main/java/com/amplifyframework/rx/RxAuthBinding.java [hunk 2]
```diff
@@ -82,6 +83,12 @@
             delegate.confirmSignUp(username, confirmationCode, onResult, onError));
     }
 
+    @Override
+    public Single<AuthSignUpResult> resendSignUpCode(
+            @NonNull String username, @NonNull AuthResendSignUpCodeOptions options) {
+        return toSingle((onResult, onError) -> delegate.resendSignUpCode(username, onResult, onError));
```
Comment (Jordan-Nelson) [inline rxbindings/src/main/java/com/amplifyframework/rx/RxAuthBinding.java:89]:
```suggestion
        return toSingle((onResult, onError) -> delegate.resendSignUpCode(username, options, onResult, onError));
```

**Comment by frimfram** [comment]
Can you please elaborate on what type of testings you have done so that this PR does not introduce regressions?  I see some unit tests are done but we are looking for more testing.

**Comment by Jordan-Nelson** [comment]
Hi @frimfram - For context, I am adding this for use in amplify-flutter.

I have tested an end to end resend sign up code flow from a flutter app with the changes in https://github.com/aws-amplify/amplify-flutter/pull/738 and confirmed that I can see the metadata in a lambda trigger.

If there is something specific you are looking for from a testing perspective, please let me know.

**Comment by frimfram** [comment]
Please make sure to test edge cases.  These are some:
- Backend error handling (list all possible error cases and test each)
- Client error such as the phone having no network or slow network.
- some other edge cases

**Comment by Jordan-Nelson** [comment]
@frimfram - I have tested the following edge cases. I first tested them with amplify android version 1.21.0 to verify the existing behavior and then tested the same scenarios against this PR (with default options and with options + metadata).

If there are other scenarios you would like to see tested, please let me know.


| Scenario                  	        | Expected Result (based on v1.21.0 behavior)           	        | PR w/o metadata 	| PR w/ metadata 	|
|---------------------------	|-------------------------------------------------------	|-----------------	|-------------	|
| No Network                	        | AuthException, cause = Unable to execute HTTP request 	|  ✅             	        |  ✅            	|
| Non-existent Username     	| UserNotFoundException                                 	                 |  ✅             	        |  ✅            	|
| No Username                         | InvalidParameterException                             	                 |  ✅             	        |  ✅            	|
| Already confirmed user    	| InvalidParameterException                             	                 |  ✅             	        |  ✅            	|
| 5+ Attempts in a row      	| LimitExceededException                                	                 |  ✅             	        |  ✅            	|
| Failure in lambda trigger 	| AuthException, cause = UserLambdaValidationException  |  ✅             	        |  ✅            	|

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/AWSCognitoAuthPlugin.java [hunk 3]
```diff
@@ -350,10 +352,17 @@ public void confirmSignUp(
     @Override
     public void resendSignUpCode(
             @NonNull String username,
+            @NonNull AuthResendSignUpCodeOptions options,
```
Comment (frimfram) [inline aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/AWSCognitoAuthPlugin.java:355]: Isn't it possible to introduce another overloaded function instead of inserting new NonNull param in a public function?  If someone was calling this function, then they would also need to change as a result of this change.  For public SDK's it's best to minimize breaking changes.

File: core-kotlin/src/main/java/com/amplifyframework/kotlin/auth/Auth.kt [hunk 4]
```diff
@@ -86,11 +87,16 @@ interface Auth {
...
+     *                If not provided, default options will be used
      * @return A sign-up result; if the code is requested, typically the result will
      *         include a next step requiring confirmation of the re-sent code.
      */
     @Throws(AuthException::class)
-    suspend fun resendSignUpCode(username: String): AuthSignUpResult
+    suspend fun resendSignUpCode(
+        username: String,
+        options: AuthResendSignUpCodeOptions = AuthResendSignUpCodeOptions.defaults()
+    ): AuthSignUpResult
```
Comment (frimfram) [inline core-kotlin/src/main/java/com/amplifyframework/kotlin/auth/Auth.kt:99]: Same concern as the above - can you think of a way to introduce this in a backward compatible way?

**Review by frimfram — CHANGES_REQUESTED** [review]

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/AWSCognitoAuthPlugin.java (hunk 3 above)
Comment (Jordan-Nelson) [reply aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/AWSCognitoAuthPlugin.java:355]: This PR does introduce a new overloaded function. The diff here is a possibly a little misleading since it shows `resendSignUpCode` being modified. `resendSignUpCode` definied on like 381 below has the same method signature as before, and it calls this newly added method with a default option. This will not be a breaking change.

File: core-kotlin/src/main/java/com/amplifyframework/kotlin/auth/Auth.kt (hunk 4 above)
Comment (Jordan-Nelson) [reply core-kotlin/src/main/java/com/amplifyframework/kotlin/auth/Auth.kt:99]: The newly added parameter has a default option. This is backwards compatible.

**Review by raphkim — APPROVED** [review]

**Review by frimfram — APPROVED** [review]
